PythonVisualizations/QuadTree.py

```diff
@@ -0,0 +1,227 @@
+                if i == self.__root:
+                    self.rootOutline = self.canvas.create_oval((i.x - self.CIRCLE_DIMEN//2-2, i.y- self.CIRCLE_DIMEN//2-2, 
+                                             i.x + self.CIRCLE_DIMEN//2+2, i.y + self.CIRCLE_DIMEN//2 +2), 
+                                            outline = "yellow", width = 3)
+                    self.canvas.lift(self.rootOutline)
+                horiz = self.canvas.create_line(i.horizontal_line, fill = self.LINE_COLOR)
+                vert = self.canvas.create_line(i.vertical_line, fill = self.LINE_COLOR)
+                self.canvas.lower(horiz), self.canvas.lower(vert)
+                self.lines.append(horiz), self.lines.append(vert)
+        
+    
+    #wrapper method for insert   
+    def insert(self, x, y, d): 
+        self.__root = self.__insert(self.__root, x, y, d)
+    
+    #creates a new node if none at designated coords
+    #otherwise adds data to existing coordinate
+    def __insert(self, n, x, y, d):
+        callEnviron = self.createCallEnvironment()
+        # return a new Node if we've reached None
+        x = int(x)
+        y = int(y)
+        if not n: 
+            node = Node(x, y, d)
+            self.nodes.append(node)
+            node.countSpaces = 1 
+                
+            
+            if self.showTree == True:
+                hor, ver = self.drawLine(node,self.parent, self.direction)
+                self.lines.append(hor), self.lines.append(ver)
+                if not self.__root: self.rootOutline = self.canvas.create_oval((x - self.CIRCLE_DIMEN//2-2, y- self.CIRCLE_DIMEN//2-2, x + self.CIRCLE_DIMEN//2+2, y + self.CIRCLE_DIMEN//2 +2), outline = "yellow", width = 3)
+            else: self.drawLine(node,self.parent, self.direction)
+            
+            oval = self.canvas.create_oval(x - self.CIRCLE_DIMEN//2, y- self.CIRCLE_DIMEN//2, x + self.CIRCLE_DIMEN//2, y + self.CIRCLE_DIMEN//2, fill = "BLACK")
+            text  = self.canvas.create_text(x-15, y - 12, text = d, fill = "red", font = ("Helvetica", "10"))
+            self.points.append(oval)
+            node.dataObjects = [text]
+      
+            self.COUNTER +=1 #keeps track of number of nodes inserted
+            
+            handler = lambda e: self.setArguments(str(x), str(y), str(d))
+            self.canvas.tag_bind(oval, '<Button>', handler)            
+            
+            return node
+        
+        # if the point to be inserted is identical to the current node, 
+        # add the data to the list of data, but don't recurse any further
+        if n.x == x and n.y == y:
+            n.data.append(d)
+            n.data.append(d)
+            key = d
+            text = self.canvas.create_text((x-3 + n.countSpaces), y - 12, fill = "red",  text = ",", font = ("Helvetica", "10"))
+            n.countSpaces+= 25
+            text  = self.canvas.create_text((x-15 + n.countSpaces), y - 12, fill = "red",  text = key, font = ("Helvetica", "10"))
```

Comment:
Does the data really need to be appended twice in lines 144 & 145?

Instead of creating new canvas text items and keeping track of counting spaces, I think it would be better to find the canvas text item and change it to be list of labels for the node.  For example, something like `self.canvas.itemconfigure(n.label, 'text', ', '.join(n.data))` assuming `n.label` holds the canvas text item ID or tag.  Also the `key` variable seems to just be a longer name for `d`.